extern "C"

mitinarseny opened the pull request:
Changed lib type to shared
Added C code example

Astrarog committed:
fix: enabled library in C code
Changed lib type to shared
Added C code example

Astrarog committed:
feat: cleaned dependences with cstdlib malloc
Implemented sqeared congruential PRNG

Astrarog committed:
fix: Unsigned random type

mitinarseny committed:
feat: clang-tidy

mitinarseny committed:
feat: c linkage

mitinarseny committed:
feat: use clang in Dockerfile

mitinarseny committed:
feat: docker clang-tidy

Astrarog committed:
fix: PAGE_SIZE initialization before main
fix: deleted standart malloc dependecies
fix: C linkage with static library

mitinarseny committed:
feat: static library by default, dockerfile

mitinarseny committed:
Merge branch 'c' of github.com:hse-malloc/malloc into c

mitinarseny committed:
Merge branch 'c' into cc

mitinarseny committed:
refactor: no noexcept in malloc and free

mitinarseny committed:
feat: hse:: -> std::

Astrarog committed:
fix: malloc declaration with noexcept for unix systems

Astrarog committed:
feat: calloc implementation
feat: new and delete c++ opetarion
note: added debug info

mitinarseny commented on src/malloc.cpp:
```diff
@@ -3,24 +3,72 @@
     extern "C" {
-        void* malloc(size_t size) {
+        void* malloc(size_t size) noexcept {
+            write(1,"MALLOC\n", 7);
```
а можно сделать, чтобы это писалось только при сборке с `HSE_MALLOC_DEBUG`?
https://stackoverflow.com/a/51190178

mitinarseny commented on src/malloc.cpp:
```diff
@@ -3,24 +3,72 @@
 #include <cstdint>
+#include <system_error>
+
+#include "unistd.h"
```
в `.cpp` файлах лучше не использовать сишные заголовки
надо посмотреть, есть ли какой-то более универсальный способ, чтобы писать логи в DEBUG режиме, а если нет - то `#include <iostream>` в случае если `#ifdef HSE_MALLOC_DEBUG` или типа того

Astrarog replied:
любые потоки используют malloc в своей реализации, поэтому у нас только системные вызовы остаются

Astrarog replied:
Фигню сказал, всё хорошо работает с std::cerr

Astrarog replied:
ну как хорошо, сегфолты ловятся не тестах....
поэтому пока только с системным вызовом

Astrarog committed:
feat: debug info only in Debug build

Astrarog committed:
feat: realloc

mitinarseny committed:
refactor: realloc

mitinarseny committed:
feat: cmake lint-fix target

mitinarseny committed:
format: lint

Astrarog committed:
feat: PRNG inital value from "/dev/random"
feat: simple thread safety

Astrarog commented:
Мне кажется надо мерджить рабочую версию и открывать новую ветку: сборку на Си мы сделали

Astrarog committed:
fix: conditional std::random_device

Astrarog committed:
fix: realloc with nullptr

Astrarog committed:
feat: llvm coverage analysis

Astrarog committed:
feat: experimental aligned_alloc

Astrarog committed:
refactor: aligned_alloc
fix: aligned_alloc with newly created chunks
feat: aligment-aware c++ new and delete operators

mitinarseny commented on src/malloc.cpp:
```diff
@@ -13,7 +13,7 @@
 
 namespace std {
-static hse::memory::Allocator _allocator{};
+thread_local hse::memory::Allocator _allocator{};
```
почему thread_local? тогда у нас каждый поток [будет](https://en.cppreference.com/w/cpp/language/storage_duration) иметь свой `_allocator`:
> **thread** storage duration. The storage for the object is allocated when the thread begins and deallocated when the thread ends. Each thread has its own instance of the object.

Astrarog replied:
а как тогда ещё достичь thread-safety?

mitinarseny replied:
с помощью мьютексов или атомиков

mitinarseny commented on README.md:
````diff
@@ -30,11 +30,23 @@ $ cmake --install build
+
+```sh
+$ cd build/test
+$ ./malloc_test 1>/dev/null 2>/dev/null && \
````
давай будем использовать [`ctest`](https://cmake.org/cmake/help/latest/manual/ctest.1.html)?
```
$ cd build
$ ctest
```

Astrarog replied:
lllvm статистика не хочет работать с ctest
там надо напрямую запускать программу
если у тебя получиться прикрутить, то будет супер

mitinarseny commented on src/memory/allocator.cpp:
```diff
@@ -79,6 +79,60 @@ MemoryControlBlock *Allocator::allocChunk(std::size_t size) {
+    // TO DO: we need new function that will guarantee allocates new Chunk with given aligment
+}
+
+std::size_t needed_shift(MemoryControlBlock* mcb, std::size_t aligment)
```
1) мы используем `camelCase` для названий переменных и функций
2) название функции нужно поменять на что-то более понятное, потому что сейчас по этой сигнатуре непонятно, что она делает. какое смещение она возвращает? почему `needed`? смещение относительно чего?

Astrarog replied:
это старый вариант смотри зарефактореный код

mitinarseny commented on src/memory/allocator.cpp:
```diff
@@ -79,6 +79,60 @@ MemoryControlBlock *Allocator::allocChunk(std::size_t size) {
+
+std::size_t needed_shift(MemoryControlBlock* mcb, std::size_t aligment)
+{
+    return (mcb->data()%aligment == 0)? 0:(aligment - (mcb->data()%aligment));
```
ты тут два раза считаешь `(mcb->data() % alignment)`

Astrarog replied:
мне кажется компилятор должен это оптимизировать

mitinarseny replied:
да, но на это лучше все-таки не рассчитывать

mitinarseny commented on src/memory/allocator.cpp:
```diff
@@ -79,6 +79,60 @@ MemoryControlBlock *Allocator::allocChunk(std::size_t size) {
+std::uintptr_t Allocator::aligned_alloc(std::size_t aligment, std::size_t size)
+{
+    MemoryControlBlock *mcb;
+    for (mcb = this->firstFree;
```
почти то же самое есть в `Allocator::searchFit()`, надо переиспользовать тот код

mitinarseny commented on src/memory/allocator.cpp:
```diff
@@ -79,6 +79,60 @@ MemoryControlBlock *Allocator::allocChunk(std::size_t size) {
+
+         mcb = mcb->nextFree()) ;
+
+    bool isNewChunk = false;
```
это зачем?

Astrarog replied:
это тоже артефакт старого кода

mitinarseny commented on src/memory/allocator.cpp:
```diff
@@ -79,6 +79,60 @@ MemoryControlBlock *Allocator::allocChunk(std::size_t size) {
+    // in order not to leak memory we need to allocace another block
+
+    // check if we have enough space to insert mcb
+    if(MemoryControlBlock::spaceNeeded(needed_shift(mcb, aligment) + size <= mcb->size()))
```
используй `mcb->fits()`

Astrarog replied:
хорошее замечание

mitinarseny commented on src/memory/allocator.h:
```diff
@@ -37,6 +37,10 @@ class Allocator {
 
+    // TO DO
+    // allocChunkAligned allocates memory pages for new block with given size and aligment
+    MemoryControlBlock *allocChunkAligned(std::size_t, std::size_t);
```
по сигнатуре функции непонятно, в каком из аргументом мы передаем `alignment`, а в каком - `size`

Astrarog replied:
можно использовать std::align_val_t, но тип тогда придётся делать лишний static_cast в самом начале
в самой стандартной библиотеке плюсов они ввели новый тип лишь для того, чтобы не было дублирования сигнатур

mitinarseny commented on src/memory/memory_control_block.h:
```diff
@@ -116,6 +116,13 @@ class MemoryControlBlock {
     void makeEndOfChunk() noexcept;
+
+    // shifts the current mcb forward if it is free
+    // it is not possible to shift the first MCB
```
первый MCB в chunk'е можно сместить:
1) если хватает места, чтобы перед ним вставить еще один непустой MCB - вставляем
2) если не хватет - просто сдвигаем, потери в памяти максимум `sizeof(MemoryControlBlock) + 1`

Astrarog replied:
можно, именно это я и делаю в aligned_alloc, если нашёл первый блок в чанке
потери памяти хочется избежать совсем, потому что чем больше чанков, тем больше будут потери памяти
однако, если unmap их так же освобождает, то возможно ими и можно пренебречь

mitinarseny commented on src/memory/memory_control_block.cpp:
```diff
@@ -133,4 +134,23 @@ void MemoryControlBlock::makeEndOfChunk() noexcept {
+
+MemoryControlBlock* MemoryControlBlock::shiftForward(std::size_t size) noexcept
+{
+    if(this->busy() || this->prev()==nullptr || size == 0)
```
1) используй `mcb->firstInChunk()`
2) надо уметь сдвигать первый блок

Astrarog replied:
хорошее замечание

mitinarseny commented on src/memory/memory_control_block.cpp:
```diff
@@ -133,4 +134,23 @@ void MemoryControlBlock::makeEndOfChunk() noexcept {
 }
+
+
+MemoryControlBlock* MemoryControlBlock::shiftForward(std::size_t size) noexcept
```
давай заменим `size` на `shift`

mitinarseny replied:
данная функция плоха тем, что ее возвращаемое значение надо не забыть приравнять к указателю на MCB, с которого мы ее вызывали:
```cpp
mcb = mcb->shiftForward(size)
```
Давай сделаем это статическим методом:
```cpp
static MemoryControlBlock::shiftForward(MemoryControlBlock *&mcb, std::size_t shift) noexcept;
```
я только не помню, можно ли писать `*&`, если нет, то `**`
она будет менять значение указателя на новый адрес смещенного MCB

Astrarog replied:
окей, да, ссылку на указатель сделать можно

mitinarseny commented on src/memory/memory_control_block.cpp:
```diff
@@ -133,4 +134,23 @@ void MemoryControlBlock::makeEndOfChunk() noexcept {
+    auto* new_mcb = reinterpret_cast<MemoryControlBlock*>(reinterpret_cast<std::uintptr_t>(this+size));
+
+    this->prev()->grow(size);
+    this->next()->setPrev(new_mcb);
```
а если это последний блок в chunk'е? `*(this->next())` будет вести к SEGFAULT'у

mitinarseny commented on src/memory/memory_control_block.cpp:
```diff
@@ -133,4 +134,23 @@ void MemoryControlBlock::makeEndOfChunk() noexcept {
+
+    new_mcb->setPrevFree(this->prevFree_);
+    new_mcb->setNextFree(this->nextFree_);
+    new_mcb->size_ = this->size_;
```
используй `mcb->setSize()` и `mcb->setFree()`

Astrarog replied:
хотел в одну операцию завернуть

mitinarseny commented on src/memory/memory_control_block.cpp:
```diff
@@ -133,4 +134,23 @@ void MemoryControlBlock::makeEndOfChunk() noexcept {
+        return this;
+
+    size = math::roundUp<std::size_t>(size, 2);
+    auto* new_mcb = reinterpret_cast<MemoryControlBlock*>(reinterpret_cast<std::uintptr_t>(this+size));
```
`newMCB`

mitinarseny commented on src/malloc.cpp:
```diff
@@ -65,6 +66,19 @@ void *realloc(void *ptr, size_t size) noexcept {
+
+void *aligned_alloc(size_t alignment, size_t size) noexcept {
+    DEBUG_LOG("ALIGNED_ALLOC");
+    if (!alignment || !size)
```
давай явно сравнивать с нулем, это хороший тон

Astrarog replied:
хорошо

mitinarseny commented on src/malloc.cpp:
```diff
@@ -80,6 +94,18 @@ void *operator new(std::size_t size) {
     }
 }
 
+void *operator new(std::size_t size, std::align_val_t al) {
```
давай `al` переименуем в `alignment`

Astrarog replied:
брал из сигнатуры стандартной библиотеки, но не вижу никаких проблем

mitinarseny commented on src/malloc.cpp:
```diff
@@ -80,6 +94,18 @@ void *operator new(std::size_t size) {
+    DEBUG_LOG("NEW");
+    try {
+        return reinterpret_cast<void *>(
+            std::_allocator.aligned_alloc(static_cast<std::size_t>(al), size));
```
почему бы `aligned_alloc` тоже не принимать параметр типа `std::align_val_t`?

Astrarog replied:
а потому что сигнатура будет отличаться, у std::align_val_t нет неявного приведения в std::size_t, поэтому не будет совместимости с Си

mitinarseny commented on src/malloc.cpp:
```diff
@@ -90,3 +116,24 @@ void operator delete(void *ptr) noexcept {
+
+void operator delete(void *ptr, std::size_t) noexcept {
+    DEBUG_LOG("DELETE");
+    if (!ptr)
```
давай делать явное сравнение с `nullptr`

Astrarog replied:
+

mitinarseny commented on src/malloc.cpp:
```diff
@@ -90,3 +116,24 @@ void operator delete(void *ptr) noexcept {
+
+void operator delete(void *ptr, std::align_val_t) noexcept {
+    DEBUG_LOG("DELETE");
+    if (!ptr)
```
давай делать явное сравнение с `nullptr`

mitinarseny commented on src/memory/allocator.cpp:
```diff
@@ -79,58 +79,57 @@ MemoryControlBlock *Allocator::allocChunk(std::size_t size) {
-}
 
-std::size_t needed_shift(MemoryControlBlock* mcb, std::size_t aligment)
+std::size_t Allocator::randomShift(MemoryControlBlock* mcb, std::size_t alignment, std::size_t needed_size) const noexcept
```
по сигнатуре функции непонятно, что она делает

mitinarseny commented on src/memory/memory_control_block.cpp:
```diff
@@ -135,6 +135,21 @@ void MemoryControlBlock::makeEndOfChunk() noexcept {
 
+bool MemoryControlBlock::isAligned(std::size_t aligment) const noexcept { return (aligment==0? false : this->data() % aligment); }
+
+std::size_t MemoryControlBlock::minNeededShift(std::size_t alignment) const noexcept
```
по сигнатуре функции непонятно, что она делает. минимальный сдвиг куда? относительно чего? почему alignment не `std::align_val_t`?

mitinarseny commented on src/memory/memory_control_block.cpp:
```diff
@@ -135,6 +135,21 @@ void MemoryControlBlock::makeEndOfChunk() noexcept {
+    return (this->isAligned(alignment)? 0:(alignment - (this->data()%alignment)));
+}
+
+std::size_t MemoryControlBlock::maxPossibleShift(std::size_t alignment, std::size_t needed_size) const noexcept
```
* по сигнатуре функции непонятно, что она делает. максимальный сдвиг куда? относительно чего? почему alignment не std::align_val_t?
* `needed_size`  -> `size`

mitinarseny commented on src/memory/memory_control_block.cpp:
```diff
@@ -135,6 +135,21 @@ void MemoryControlBlock::makeEndOfChunk() noexcept {
+std::size_t MemoryControlBlock::maxPossibleShift(std::size_t alignment, std::size_t needed_size) const noexcept
+{
+    std::size_t max_shift = this->minNeededShift(alignment);
+    if(needed_size + max_shift > this->size()) // it is not possible
```
`mcb->fits()`

mitinarseny commented on src/memory/memory_control_block.h:
```diff
@@ -117,6 +117,17 @@ class MemoryControlBlock {
+    bool isAligned(std::size_t) const noexcept;
+
+    // returns the minimum needed shift for current block to be aligned
+    std::size_t minNeededShift(std::size_t) const noexcept;
```
по сигнатуре функции непонятно, что она делает. минимальный сдвиг куда? относительно чего? почему alignment не std::align_val_t?

mitinarseny commented on src/memory/memory_control_block.h:
```diff
@@ -117,6 +117,17 @@ class MemoryControlBlock {
+
+    // returns the maximum possible shift for current block to store provided size and to be alinged
+    // returns zero if it is not possible
+    std::size_t maxPossibleShift(std::size_t, std::size_t) const noexcept;
```
по сигнатуре функции непонятно, что она делает. максимальный сдвиг куда? относительно чего? почему alignment не std::align_val_t?

Astrarog replied:
по всем shift-функциям:
сдвиг делается вперёд относительно mcb->data().
минимальный необходимый сдвиг - на сколько байт надо минимально сдвинуться, чтобы mcb->data() было выровнено по aligment (независимо от того, какого размера данные мы хотим поместить в в mcb
максимальный возможный сдвиг - насколько мы максимально можем сдвинуть mcb вперёд, чтобы 1) хранить size значений в mcb, 2) быть выравненным по aligment (если это возможно)

согласен, что по сигнатуре не понятно, где что, но предлагаю не aligment не хранить в std::align_val_t (так как придётся явно передавать значение в этом типе), а сделать в классе "using aligment_t = std::size_t; "

mitinarseny commented on src/memory/allocator.cpp:
```diff
@@ -79,58 +79,57 @@ MemoryControlBlock *Allocator::allocChunk(std::size_t size) {
-    // previous extist and we can resize is
-    if(mcb->prev()!=nullptr)
+    std::size_t min_shift;
+    for (mcb = this->firstFree, min_shift = mcb->minNeededShift(alignment);
```
надо обобщить `Allocator::searchFit()`, чтобы он искал блок по предикату

Astrarog replied:
да, хорошее предложение

mitinarseny commented on src/memory/allocator.cpp:
```diff
@@ -79,58 +79,57 @@ MemoryControlBlock *Allocator::allocChunk(std::size_t size) {
+                if(MemoryControlBlock::spaceNeeded(1) <= mcb->maxPossibleShift(alignment, size))
+                    mcb = mcb->split(1); // at this point mcb->prev() will not be nullptr
+                else
+                    continue;
```
сначала проверяй на противоположное условие и делай `continue`, а потом делай `mcb->split()` и `break`

mitinarseny commented on src/memory/allocator.cpp:
```diff
@@ -251,8 +251,8 @@ MemoryControlBlock *Allocator::realloc(MemoryControlBlock *mcb,
     std::copy(
-        reinterpret_cast<unsigned char *>(oldMCB->data()),
-        reinterpret_cast<unsigned char *>(oldMCB->data() + oldMCB->size()),
+        reinterpret_cast<unsigned short *>(oldMCB->data()),
```
давай тогда уж `uint8_t`

Astrarog replied:
uint16_t ты имел ввиду

mitinarseny committed:
refactor: free blocks list is in separate class

Astrarog commented on src/memory/allocator.cpp:
```diff
@@ -12,150 +12,131 @@
 
 namespace hse::memory {
 
+sc69069_t Allocator::randomGenerator(std::random_device
```
оставить thread_local

Astrarog committed:
feat: malloc on Windows with VirtualAlloc

Astrarog commented on src/system/system.cpp:
```diff
@@ -13,23 +21,44 @@ namespace hse::system {
+    val = ::sysconf(name);
     if (val == -1)
         throw std::system_error(errno, std::system_category(), "sysconf");
+#elif defined HAVE_VIRTUAL_ALLOC
```
Возможно нужно создать отдельные файлы с реализацией на разных системах типа 
system_windows.cpp

mitinarseny commented on src/system/system.cpp:
```diff
@@ -13,23 +21,44 @@ namespace hse::system {
     if (::munmap(reinterpret_cast<void *>(addr), len) == -1)
         throw std::system_error(errno, std::system_category(), "munmap");
+#elif defined HAVE_VIRTUAL_ALLOC
+    if (::VirtualFree(reinterpret_cast<void *>(addr), 0, MEM_RELEASE) == 0)
```
ты тут не учитываешь параметр `len`, который для нас важен

Astrarog replied:
Это так, но в винде сказали, что при использовании `MEM_RELEASE` второй аргумент надо выставлять в `0`. Возможно стоит копнуть глубже в этот вопрос

mitinarseny committed:
refactor: aligned_alloc

mitinarseny committed:
feat: allocChunk if there is no suitable block in alignedAllock

mitinarseny committed:
format: remove TODOs

mitinarseny committed:
feat: aligned_alloc rand

Astrarog commented on include/malloc.h:
```diff
@@ -19,11 +22,11 @@ extern "C" {
+__NODISCARD__ void *calloc(size_t count, size_t size) __NOEXCEPT__;
+__NODISCARD__ void *realloc(void *ptr, size_t size) __NOEXCEPT__;
+__NODISCARD__ void *aligned_alloc(size_t alignment, size_t size) __NOEXCEPT__;
 void free(void *) __NOEXCEPT__;
```
а free не надо пометить?

Astrarog commented on src/memory/memory_control_block.h:
```diff
@@ -64,6 +64,10 @@ class MemoryControlBlock {
 
+    // rsplit is the same as split, but it makes second (right)
+    // block to be of given size
+    MemoryControlBlock* rsplit(std::size_t) noexcept;
```
Полезная функция!

Astrarog committed:
feat: Using catch2 in CMake to run tests

Astrarog committed:
reafactor: restructured project building process to properly test implemented funtions

Astrarog committed:
feat: catch2 simple test

mitinarseny committed:
feat: setBusy sets busy bit to given one

mitinarseny committed:
Merge branch 'c' of github.com:hse-malloc/malloc into c

Astrarog committed:
feat: unit tests for malloc, calloc, realloc, aligned_alloc

mitinarseny committed:
fix: git in docker to clone catch2, use concepts

mitinarseny committed:
Merge branch 'c' of github.com:hse-malloc/malloc into c

mitinarseny committed:
refactor: malloc-implementation -> hse_malloc

Astrarog committed:
fix: catch2 without installation in system

mitinarseny committed:
fix: split and shiftForward should mantain one-sequential-free-block rule

mitinarseny committed:
refactor: tests use std::span

Astrarog committed:
fix: using std::array for big objects in tests

Astrarog committed:
fix: realloc smaller after big malloc

Astrarog committed:
feat: test for multiple usage of malloc, aligned_alloc and free

Astrarog committed:
feat: disable randomisation with HSE_MALLOC_NO_RANDOM

Astrarog committed:
fix: ability to use HSE_MALLOC_NO_RANDOM via cmake

Astrarog committed:
Update README.md

Astrarog committed:
feat: enable testing only in Debug build

Commit:
refactor: MCB::next() and MCB::data() are static due to this can be a pointer on stack

Commit:
format: lint

Commit:
refactor: do not use this as pointer to MCB as it can be on stack

Commit:
format: lint
